rust_icu_uloc/src/lib.rs

```diff
@@ -346,43 +496,95 @@ pub fn accept_language(
         .map(|uloc| (Some(uloc), accept_result))
 }
 
+/// Implements `uloc_toUnicodeLocaleKey` from ICU4C.
+pub fn to_unicode_locale_key(legacy_keyword: &str) -> Option<String> {
+    let legacy_keyword = str_to_cstring(legacy_keyword).unwrap();
+    let unicode_keyword: Option<ffi::CString> = unsafe {
+        let ptr = versioned_function!(uloc_toUnicodeLocaleKey)(legacy_keyword.as_ptr());
+        ptr.as_ref().map(|ptr| ffi::CStr::from_ptr(ptr).to_owned())
+    };
+    unicode_keyword.map(|cstring| cstring_to_string(&cstring).unwrap())
+}
+
+/// Implements `uloc_toUnicodeLocaleType` from ICU4C.
+pub fn to_unicode_locale_type(legacy_keyword: &str, legacy_value: &str) -> Option<String> {
+    let legacy_keyword = str_to_cstring(legacy_keyword).unwrap();
+    let legacy_value = str_to_cstring(legacy_value).unwrap();
+    let unicode_value: Option<ffi::CString> = unsafe {
+        let ptr = versioned_function!(uloc_toUnicodeLocaleType)(
+            legacy_keyword.as_ptr(),
+            legacy_value.as_ptr(),
+        );
+        ptr.as_ref().map(|ptr| ffi::CStr::from_ptr(ptr).to_owned())
+    };
+    unicode_value.map(|cstring| cstring_to_string(&cstring).unwrap())
+}
+
+/// Implements `uloc_toLegacyKey` from ICU4C.
+pub fn to_legacy_key(unicode_keyword: &str) -> Option<String> {
+    let unicode_keyword = str_to_cstring(unicode_keyword).unwrap();
+    let legacy_keyword: Option<ffi::CString> = unsafe {
+        let ptr = versioned_function!(uloc_toLegacyKey)(unicode_keyword.as_ptr());
+        ptr.as_ref().map(|ptr| ffi::CStr::from_ptr(ptr).to_owned())
+    };
+    legacy_keyword.map(|cstring| cstring_to_string(&cstring).unwrap())
+}
+
+fn str_to_cstring(input: &str) -> Result<ffi::CString, common::Error> {
+    ffi::CString::new(input).map_err(|_| common::Error::string_with_interior_nul())
+}
+
+fn cstring_to_string(input: &ffi::CString) -> Result<String, common::Error> {
```

Comment:
If we know in advance that input is a valid UTF-8 string (which we could since this is a private function), we could use the same trick as above to simplify the function signature to be just `Result<String>`.

If we don't, then looks like this should return a Result<String, Utf8Error> -- Looks to me like we're doing `unwrap` on the result anyways, so no need to convert it before discarding.

Reply:
You're right, all possible inputs to this function are coming from ICU data, so we can safely assume they're valid UTF-8.